### ISSUE TITLE
test(features): single-language legs fail to build tests/api/ast_seam_test.rs

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#1426]
## Summary

**Fixed** in `test(features): build tests/api under single-language feature
sets`.

Two single-language feature subsets were reported failing on `main`, both in
`tests/api/ast_seam_test.rs` rather than in the library — but at *different*
stages, which matters for where a gate would have to sit:

```console
$ cargo check --no-default-features --features c -p big-code-analysis --all-targets
error[E0433]: cannot find type `PathBuf` in this scope
error[E0433]: cannot find type `MetricsOptions` in this scope
error: could not compile `big-code-analysis` (test "api") due to 2 previous errors

$ cargo clippy --no-default-features --features tcl -p big-code-analysis --all-targets -- -D warnings
error: for loop over a single element
error: could not compile `big-code-analysis` (test "api") due to 1 previous error
```

- `--features c` is a genuine **rustc** failure: unresolved `PathBuf` /
  `MetricsOptions` imports, reachable only once the `rust`-gated block is
  compiled out.
- `--features tcl` **compiles clean** (`cargo check` passes) and fails only
  under `clippy -D warnings`: `clippy::single_element_loop`, from a loop whose
  row table reduces to one element when the other languages are gated out.

Both reproduced directly on the current tree, not inferred.

Found while fixing #1286 / #1411. Verified pre-existing by stashing that work
and rebuilding, so it is untouched by that change and predates this batch.

## What the full sweep found

Sweeping **all 22** single-language subsets — not just the two above — under
`cargo clippy --no-default-features --features <lang> -p big-code-analysis
--all-targets -- -D warnings` gave **5 failing legs from 3 distinct defects**:

| Leg(s) | Defect | Diagnostic |
|---|---|---|
| `c` | `PathBuf` / `MetricsOptions` unions too **narrow** — gated `any(rust, python, cpp)`, used by a `c`-gated test | `E0433` |
| `cpp`, `python` | `ops_tree_contains_operator` union too **wide** — gated `any(rust, python, cpp)`, one `rust`-gated caller | `dead_code` |
| `tcl`, `irules` | two-row Tcl-family table with cfg-gated rows collapses to a singleton | `clippy::single_element_loop` |

The `cpp` / `python` class was not in the original report. It is the exact
mirror of the `c` class: the same hand-maintained union, wrong in the other
direction.

`big-code-analysis-ast` is clean on all 22 single-language legs, before and
after.

## Why CI could not see it

The `feature-matrix` job in `.github/workflows/ci.yml` covered only
`default`, `no-default-features`, and `minimal-langs` (`rust,typescript`) for
the lib and ast crates, plus `default` / `no-default-features` for cli and web.
No leg built a *single* non-`rust` language — and because `rust` is in
`minimal-langs`, nearly every union in the file was satisfied by accident.

This matters more than a normal gap, because `--features go` and
`--features tcl` are the reproducer subsets `.claude/rules/testing.md`
prescribes for verifying that a feature-gated fixture table is correctly
gated ("Verify against a subset that enables **none** of the named features…
a single non-listed language (`--features go`) is the reproducer"). A
contributor following the rule hit a compile error in an unrelated file and
had no way to tell it from a defect in their own change — the same
false-attribution failure #1220 / PR #1221 and #1286 are about, one layer
down.

## What shipped

**1. `tests/api/ast_seam_test.rs`**

- `use std::path::PathBuf` and `use big_code_analysis::MetricsOptions` gain
  `feature = "c"`.
- `ops_tree_contains_operator` narrows from `any(rust, python, cpp)` to
  `feature = "rust"`, its sole caller's gate.
- `count_string_agrees_with_find_on_tcl_family_bodies` drops the cfg-gated
  row table for two `#[cfg]` blocks calling a shared local `check`, matching
  the sibling test directly above it. Correct at one row or two. The
  `#[cfg(any(tcl, irules))]` gate on the `fn` and the `ran > 0` non-vacuity
  assertion both stay, per `.claude/rules/testing.md`.

**2. `.github/workflows/ci.yml`, `feature-matrix`**

- `cargo check --all-targets` → `cargo clippy --all-targets -- -D warnings`.
  Of the three classes, `check` catches one: `dead_code` is a rustc lint it
  reports as a warning and does not promote, and `single_element_loop` is a
  clippy lint it never runs. All ten pre-existing legs were verified against
  the stricter command before the promotion landed.
- New legs: `c`, `go`, `python`, `tcl` (lib) and `go` (ast).

On the leg choice: the original suggestion of `go` alone would have caught
**none** of the three defects — `go` enables no per-language test in the file,
so it is close to redundant with the existing `no-default-features` leg for
this purpose. It is kept because it is the subset the testing rule names and
it does exercise one language's own library code, but the legs that
discriminate are `c`, `python` and `tcl`, each of which failed on `main` and
passes now.

## Verification

- The same 22-leg sweep re-run after the fix: **22 pass, 0 fail** (17/5
  before). Same script both times, and the before run was not uniform — three
  distinct error texts — so the harness discriminates.
- `count_string_agrees_with_find_on_tcl_family_bodies` and its sibling *run*
  and pass under `--features tcl`, `--features irules`, and `--all-features`,
  not merely compile.
- `make pre-commit`: `BCA_GATE: pass (gate=pre-commit)`.
- `make actionlint` clean.

## Scope note

This was a *compile* failure, distinct from #1285 / #1413, which are about
tests that compile fine and panic at run time under a partial feature set.
Fixing this does not advance those. Corroborating datum for them: running
`cargo test --no-default-features --features c -p big-code-analysis --test api`
now that it compiles gives **24 run-time failures**, all in
`suppression_test.rs` (20) and `derive_eq_hash_ord.rs` (4), which parse Rust
fixtures with no `rust` feature gate. `ast_seam_test` itself is clean on that
leg. Untouched here; that is #1285's job.

**Comment by dekobon** [comment]
Fixed on `fix/1426-single-language-feature-legs`.

Research note worth recording separately from the body: the report named two
failing legs; a full sweep of all 22 single-language subsets found **five**,
from **three** defects. The missing class was `cpp` / `python` —
`ops_tree_contains_operator` gated on `any(rust, python, cpp)` with a single
`rust`-gated caller, so it builds *unused* and `-D warnings` promotes the
`dead_code` to an error. That is the exact mirror of the reported `c` failure:
the same hand-maintained cfg union, wrong in the opposite direction. Neither
direction is visible from reading the file — only from building a subset that
enables one language and not the others.

Two consequences for the gate design, both of which changed what shipped:

1. **`go` alone would not have caught any of the three.** It enables no
   per-language test in `ast_seam_test.rs`, so for this file it is close to
   redundant with the existing `no-default-features` leg. Kept (it is the
   subset `.claude/rules/testing.md` names, and it does compile one language's
   own library code), but the discriminating legs are `c`, `python` and `tcl`.
2. **`cargo check` catches one class of three.** `dead_code` is a rustc lint
   `check` reports as a warning and never promotes; `single_element_loop` is a
   clippy lint `check` does not run. So the whole `feature-matrix` job moved to
   `cargo clippy --all-targets -- -D warnings`, which also upgrades the ten
   pre-existing legs — each verified passing under the stricter command before
   the promotion landed, so no unrelated failure rode in with it.

`big-code-analysis-ast` is clean on all 22 single-language legs both before and
after, so the `go (ast)` leg gates nothing broken today; it is there to keep
the lib/ast mirror the matrix already maintains.

**Comment by dekobon** [comment]
Follow-up after code review — a fourth defect and two corrections to what I
wrote above. The commit and the issue body are updated.

**A fourth instance of the same drift, and the reason the sweep missed it.**
`use big_code_analysis::{LANG, Source}` had the identical `c`-shaped hole as
`PathBuf`, but the 22-leg sweep reported `--features c` green, because that
union ends in `not(feature = "javascript")`:

```rust
#[cfg(any(rust, python, c, cpp, tcl, irules, not(feature = "javascript")))]
use big_code_analysis::{LANG, Source};
```

The trailing arm is true in *every* single-language leg, so it satisfies the
union by itself and none of the language arms is ever load-bearing. The arm is
genuinely needed — `ast_parse_returns_language_disabled_for_off_feature` names
`LANG::Javascript` and `Source::new` with no language enabled at all — so it
cannot just be deleted. The consequence is that a whole-matrix-green result
said nothing about that union.

Confirmed by perturbation rather than inspection: removing `feature = "c"`
from the arm list reproduces both `E0433`s under
`--features c,javascript`, and restores clean. That is also why the matrix now
carries one two-language leg (`c,javascript`) — otherwise the fix to that
union would ship with no gate at all, which AGENTS.md's
"add a regression test that would catch the exact bug if reintroduced" rules
out.

**Correction 1 — `cargo check` would have caught two of three, not one.**
`ci.yml` sets `RUSTFLAGS: "-D warnings"` at the workflow level, so `cargo
check` in this job already promoted the `dead_code` class. Only
`clippy::single_element_loop` is genuinely out of `check`'s reach. The move to
clippy is still right, but I overstated its value; the comment and CHANGELOG
now say what is actually true, and the explicit `-- -D warnings` is documented
as insurance against that env var being narrowed later.

**Correction 2 — `go` does enable per-language code.** I wrote that `go`
"enables no per-language test in the lib suite". False: there are thirteen
`feature = "go"` gates under `src/`, including
`go_emits_npa_and_npm_but_never_wmc` in `src/metrics/container_scope_tests.rs`.
The accurate reason to keep the leg is narrower — `go` enables none of the
five features the *shared items in `ast_seam_test.rs`* are gated on, so it is
the "outside every union" shape.
